lib/routing_nb.h

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2018        Vmware
+ *                           Vishal Dhingra
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+#ifndef _FRR_ROUTING_NB_H_
+#define _FRR_ROUTING_NB_H_
+
+extern const struct frr_yang_module_info frr_routing_info;
+
+/* Mandatory callbacks. */
+int routing_control_plane_protocols_control_plane_protocol_create(
+	struct nb_cb_create_args *args);
+int routing_control_plane_protocols_control_plane_protocol_destroy(
+	struct nb_cb_destroy_args *args);
+
+#define FRR_ROUTING_XPATH                                                      \
+	"/frr-routing:routing/control-plane-protocols/control-plane-protocol"
+
+#define FRR_ROUTING_KEY_XPATH                                                  \
+	"/frr-routing:routing/control-plane-protocols/"                        \
+	"control-plane-protocol[type='%s'][name='%s'][vrf='%s']"
+
+typedef int (*routing_create_validate_cbs)(struct nb_cb_create_args *args);
+
+/*
+ * callbacks for routing to handle configuration events
+ * based on the control plane protocol
+ */
+struct routing_configuration_event_callback {
+	routing_create_validate_cbs cv_cbs;
+};
```

Comment:
nit: I think it'd be better to use our hook framework [1] to do this more elegantly.

[1] http://docs.frrouting.org/projects/dev-guide/en/latest/hooks.html

Reply:
done